config.go

```diff
@@ -0,0 +1,151 @@
+package analytics
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/segmentio/backo-go"
+	"github.com/xtgo/uuid"
+)
+
+// Instances of this type carry the different configuration options that may
+// be set when instantiating a client.
+//
+// Each field's zero-value is either meaningful or interpreted as using the
+// default value defined by the library.
+type Config struct {
+
+	// The endpoint to which the client connect and send their messages, set to
+	// `DefaultEndpoint` by default.
+	Endpoint string
+
+	// The flushing interval of the client. Messages will be sent when they've
+	// been queued up to the maximum batch size or when the flushing interval
+	// timer triggers.
+	Interval time.Duration
+
+	// The HTTP transport used by the client, this allows an application to
+	// redefine how requests are being sent at the HTTP level (for example,
+	// to change the connection pooling policy).
+	// If none is specified the client uses `http.DefaultTransport`.
+	Transport http.RoundTripper
+
+	// The logger used by the client to output info or error messages when that
+	// are generated by background operations.
+	// If none is specified the client uses a standard logger that outputs to
+	// `os.Stderr`.
+	Logger Logger
+
+	// The maximum number of messages that will be sent in one API call.
+	// Messages will be sent when they've been queued up to the maximum batch
+	// size or when the flushing interval timer triggers.
+	// Note that the API will still enfore a 500KB limit on each HTTP request
+	// which is independant from the number of embedded messages.
+	BatchSize int
```

Comment:
I tend to default to using `int` everywhere usually even for stuff that can be negative just to make it easy to integrate with most of the Go code out there that uses `int` and rarely `uint` (think functions like `len` which return `int`, you suddenly need a type conversion if you want to inject it here).

The main issue is that a `uint` cast from an integer value that is negative is totally valid but likely produce a value that wasn't the writer's intent. If we let negative values be set here at least we'll catch them at runtime and return an error instead of creating a client with a batch size of a few billion messages.

Discussion on that: https://www.reddit.com/r/golang/comments/2q5vdu/int_vs_uint/